Functions to create indication subsets and restriction subsets following a similar pattern

azimov opened the pull request.

azimov committed:
Started cohort definition set indication and restriction subset recipes

azimov committed:
Added exclude on index subset definition

azimov committed:
Merge branch 'develop' into indication-subset-recipes

azimov committed:
docstrings

azimov committed:
Merge branch 'develop' into indication-subset-recipes

azimov committed:
Doc strings tidy up

azimov committed:
Roxygen fights

azimov committed:
Gen ai unit tests

azimov committed:
GenAI tests fixed

azimov committed:
Fixed crappy errors in roxygen

azimov committed:
docstring fixes/changes

anthonysena committed:
Merge branch 'develop' into indication-subset-recipes

anthonysena commented:
@azimov nice contribution thanks for putting this together as it will help streamline the ways in which we work with subset definitions. To illustrate how this works, I've modified one of the unit tests:

To start load up one of our test cohortDefinitionSets in the package:

``` r
library(CohortGenerator)
#> Loading required package: DatabaseConnector
#> Loading required package: R6

initialSet <- getCohortDefinitionSet(
  settingsFileName = "testdata/name/Cohorts.csv",
  jsonFolder = "testdata/name/cohorts",
  sqlFolder = "testdata/name/sql/sql_server",
  cohortFileNameFormat = "%s",
  cohortFileNameValue = c("cohortName"),
  packageName = "CohortGenerator",
  verbose = FALSE
)
#> Loading cohortDefinitionSet

print(initialSet[,c("cohortId", "cohortName")])
#>   cohortId         cohortName
#> 1        1          celecoxib
#> 2        2     celecoxibAge40
#> 3        3 celecoxibAge40Male
#> 4        4  celecoxibCensored

```
Next, subset cohorts 1 & 2 by an "indication" cohort 3:

``` r
res <- addIndicationSubsetDefinition(
  cohortDefinitionSet = initialSet,
  targetCohortIds = c(1, 2),
  indicationCohortIds = c(3),
  definitionId = 10
)

print(res[,c("cohortId", "cohortName", "subsetParent", "subsetDefinitionId", "isSubset")])
#>   cohortId                                cohortName subsetParent
#> 1        1                                 celecoxib            1
#> 2        2                            celecoxibAge40            2
#> 3        3                        celecoxibAge40Male            3
#> 4        4                         celecoxibCensored            4
#> 5     1010      celecoxib - study population + first            1
#> 6     2010 celecoxibAge40 - study population + first            2
#>   subsetDefinitionId isSubset
#> 1                 NA    FALSE
#> 2                 NA    FALSE
#> 3                 NA    FALSE
#> 4                 NA    FALSE
#> 5                 10     TRUE
#> 6                 10     TRUE
```

We can then interrogate the cohortDefinitionSet to get **all** subset definitions that were created using the `addIndicationSubsetDefinition`:

```r
subsetDefinitionId <- getIndicationSubsetDefinitionIds(res)
```
We can then use this information to filter the cohortDefinitionSet to those cohorts that were defined using an indication subset definition:

```r
# Get the indication cohorts
newCohorts <- res |>
  dplyr::filter(subsetDefinitionId == subsetDefinitionId) |>
  dplyr::select(cohortId, cohortName, subsetParent, isSubset)
print(newCohorts)
#>   cohortId                                cohortName subsetParent isSubset
#> 1     1010      celecoxib - study population + first            1     TRUE
#> 2     2010 celecoxibAge40 - study population + first            2     TRUE
```

<sup>Created on 2025-09-24 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,249 @@
+#' The approach applies this definition to an exposure or set of exposures, requiring individuals to have
+#' a prior history of a condition before receiving treatment.
+#' This function aims to make parameterization of study execution explicit.
+#' Additionally, it attaches an attribute to the cohort definition set.
```
I found this description a bit confusing. I think the aim of this function is: _Create an "indication" subset whereby `targetCohortIds` are subsetted to cohort members that have any/all `indicationCohortIds` overlapping the target cohort start date. This function includes parameters to define the overlap of time and other subsetting features._

You also mention: _This function aims to make parameterization of study execution explicit._ I think you mean study design? Not sure if this help execution but I may be missing your point.

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,249 @@
+addIndicationSubsetDefinition <- function(cohortDefinitionSet,
+                                          targetCohortIds,
+                                          indicationCohortIds,
+                                          definitionId,
```
Perhaps this should be `subsetDefinitionId`?

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,249 @@
+                                          targetCohortIds,
+                                          indicationCohortIds,
+                                          definitionId,
+                                          name = "study population + first",
```
Should this be "study population + indication" vs first?

azimov replied:
Thinking about this, I think this should be a required parameter for the user

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,249 @@
+#' Add Restriction Subset Definition
+#' @export
+#' @description
+#' Utility pattern for creating cohort subset definitions as a standard approach for indicated drugs.
+#' Restriction subset definitions are twins of indication definitions. They should apply the same core properites
+#' to a base exposure cohort (i.e. study dates,  required prior observation time, ages, gender) as indications but,
+#' crucially, they do not require history of any prior condition(s).
+#'
+#' This is useful in the context of comparing drug exposure + indication population, to population as a whole.
+#'
+#' The prefered use of this function is to create this in conjunction with the target population.
```
I'd prefer if this description focused on the functionality and then later described the difference between `addRestrictuionSubsetDefinition` and `addIndicationSubsetDefinition`.

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,249 @@
+#' @param exclusionWindow                           Days Default is 0 (target index date). by changing this
+#'                                                  you can adjust the period around target index for which you would
+#'                                                  exclude members.
+addExcludeOnIndexSubsetDefinition <- function(cohortDefinitionSet,
```
Any reason why this particular function does not store information as attributes of the cohortDefinitionSet and have a related `get` function?

azimov replied:
This change has been made.

anthonysena replied:
Thanks! Looks like the example just needs some adjustment as noted for the other function.

azimov committed:
Tidy up of documentation and usage examples

azimov commented:
> @azimov nice contribution thanks for putting this together as it will help streamline the ways in which we work with subset definitions. To illustrate how this works, I've modified one of the unit tests:
> 
> To start load up one of our test cohortDefinitionSets in the package:
> 
> ```r
> library(CohortGenerator)
> #> Loading required package: DatabaseConnector
> #> Loading required package: R6
> 
> initialSet <- getCohortDefinitionSet(
>   settingsFileName = "testdata/name/Cohorts.csv",
>   jsonFolder = "testdata/name/cohorts",
>   sqlFolder = "testdata/name/sql/sql_server",
>   cohortFileNameFormat = "%s",
>   cohortFileNameValue = c("cohortName"),
>   packageName = "CohortGenerator",
>   verbose = FALSE
> )
> #> Loading cohortDefinitionSet
> 
> print(initialSet[,c("cohortId", "cohortName")])
> #>   cohortId         cohortName
> #> 1        1          celecoxib
> #> 2        2     celecoxibAge40
> #> 3        3 celecoxibAge40Male
> #> 4        4  celecoxibCensored
> ```
> 
> Next, subset cohorts 1 & 2 by an "indication" cohort 3:
> 
> ```r
> res <- addIndicationSubsetDefinition(
>   cohortDefinitionSet = initialSet,
>   targetCohortIds = c(1, 2),
>   indicationCohortIds = c(3),
>   definitionId = 10
> )
> 
> print(res[,c("cohortId", "cohortName", "subsetParent", "subsetDefinitionId", "isSubset")])
> #>   cohortId                                cohortName subsetParent
> #> 1        1                                 celecoxib            1
> #> 2        2                            celecoxibAge40            2
> #> 3        3                        celecoxibAge40Male            3
> #> 4        4                         celecoxibCensored            4
> #> 5     1010      celecoxib - study population + first            1
> #> 6     2010 celecoxibAge40 - study population + first            2
> #>   subsetDefinitionId isSubset
> #> 1                 NA    FALSE
> #> 2                 NA    FALSE
> #> 3                 NA    FALSE
> #> 4                 NA    FALSE
> #> 5                 10     TRUE
> #> 6                 10     TRUE
> ```
> 
> We can then interrogate the cohortDefinitionSet to get **all** subset definitions that were created using the `addIndicationSubsetDefinition`:
> 
> ```r
> subsetDefinitionId <- getIndicationSubsetDefinitionIds(res)
> ```
> 
> We can then use this information to filter the cohortDefinitionSet to those cohorts that were defined using an indication subset definition:
> 
> ```r
> # Get the indication cohorts
> newCohorts <- res |>
>   dplyr::filter(subsetDefinitionId == subsetDefinitionId) |>
>   dplyr::select(cohortId, cohortName, subsetParent, isSubset)
> print(newCohorts)
> #>   cohortId                                cohortName subsetParent isSubset
> #> 1     1010      celecoxib - study population + first            1     TRUE
> #> 2     2010 celecoxibAge40 - study population + first            2     TRUE
> ```
> 
> Created on 2025-09-24 with [reprex v2.1.1](https://reprex.tidyverse.org)

I have added this as example code to the definitions

azimov committed:
docstring

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,376 @@
+#'   targetCohortIds = c(1, 2),
+#'   indicationCohortIds = c(3),
+#'   subsetDefinitionId = 20
+#' )
```
The comment & function parameters need adjustment for the addRestrictionSubsetDefinition function

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,376 @@
+#'
+#' print(res[, c("cohortId", "cohortName", "subsetParent", "subsetDefinitionId", "isSubset")])
+#'
+#' # Get all subset definitions that were created using the addIndicationSubsetDefinition:
```
addRestrictionSubsetDefinition instead of addIndicationSubsetDefinition

anthonysena commented on R/SubsetRecipeFunctions.R:
```diff
@@ -0,0 +1,376 @@
+#' # Get all subset definitions that were created using the addIndicationSubsetDefinition:
+#' subsetDefinitionId <- getRestrictionSubsetDefinitionIds(res)
+#'
+#' # Filter the cohortDefinitionSet to those cohorts defined using an indication subset definition:
```
restriction instead of indication

azimov committed:
typos and clean up for documentation

anthonysena committed:
Regen docs

anthonysena committed:
Fix spelling errors and roxygen comment issue

anthonysena committed:
Fix inconsistent function name and pkgdown.yml

anthonysena committed:
Test coverage improvement

anthonysena committed:
Remove extraneous .gitignore entries

anthonysena committed:
Whitespace

azimov commented on .gitignore:
@anthonysena I don't think you want to do this. People should not check these in to the package ever - these can differ between developers and/or contain some sensitive information (e.g. if the user has to route http traffic through a proxy). We certainly don't want files in the renv folder getting checked in as this could be huge

anthonysena replied:
The files below should not exist in the package project directory:

```
.Rprofile
renv/
/renv.lock
```

It doesn't hurt to exclude them but not sure why this is necessary?

anthonysena committed:
Bug fix with ageMin/Max

anthonysena committed:
Remove default for subset def name

anthonysena committed:
Fix test
